clippy_lints/src/no_effect.rs

```diff
@@ -94,33 +94,75 @@ impl<'tcx> LateLintPass<'tcx> for NoEffect {
                         return;
                     }
-                    if let Some(snip) = snippet_opt(cx, e.span) {
-                        snippet.push_str(&snip);
-                        snippet.push(';');
-                    } else {
-                        return;
+                }
+                if is_function_used_in_index(expr) {
+                    let snippet;
+                    if_chain! {
+                        if let Some(arr) = snippet_opt(cx, reduced[0].span);
+                        if let Some(func) = snippet_opt(cx, reduced[1].span);
+                        then {
+                            snippet = format!("assert!({}.len() > {});", &arr, &func);
+                        } else {
+                            return;
+                        }
                     }
+                    span_lint_hir_and_then(
+                        cx,
+                        UNNECESSARY_OPERATION,
+                        expr.hir_id,
+                        stmt.span,
+                        "statement can be reduced",
+                        |diag| {
+                            diag.span_suggestion(stmt.span, "replace it with", snippet, Applicability::MaybeIncorrect);
+                        },
+                    );
+                } else {
+                    let mut snippet = String::new();
+                    for e in reduced {
+                        if let Some(snip) = snippet_opt(cx, e.span) {
+                            snippet.push_str(&snip);
+                            snippet.push(';');
+                        } else {
+                            return;
+                        }
+                    }
+                    span_lint_hir_and_then(
+                        cx,
+                        UNNECESSARY_OPERATION,
+                        expr.hir_id,
+                        stmt.span,
+                        "statement can be reduced",
+                        |diag| {
+                            diag.span_suggestion(
+                                stmt.span,
+                                "replace it with",
+                                snippet,
+                                Applicability::MachineApplicable,
+                            );
+                        },
+                    );
                 }
-                span_lint_hir_and_then(
-                    cx,
-                    UNNECESSARY_OPERATION,
-                    expr.hir_id,
-                    stmt.span,
-                    "statement can be reduced",
-                    |diag| {
-                        diag.span_suggestion(stmt.span, "replace it with", snippet, Applicability::MachineApplicable);
-                    },
-                );
             }
         }
     }
 }
 
+fn is_function_used_in_index(expr: &Expr<'_>) -> bool {
+    if_chain! {
+        if let ExprKind::Index(_, b) = &expr.kind;
+        if let ExprKind::Call(..) | ExprKind::MethodCall(..) = &b.kind;
```

Comment:
Does it matter that this is a call? For example a field access would lead to the same problem, wouldn't it?

So generally, should we make this exception for index operations, regardless the expression used as the index?

Reply:
Yes, you are right @flip1995 . The compiler will warn about unconditional panic if you just use hardcoded number or local struct field, but just passing struct as argument and using its field is already enough to avoid it. 

So I think it's better then to make this general to index operations.

I am not sure if I understood your comment abut error message completely.